orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1386 @@
+     * next hop.
+     */
+    if (isLabeled() &&
+        gNeighOrch->hasNextHop(m_nh_key) &&
+        (gNeighOrch->getNextHopRefCount(m_nh_key) == 0))
+    {
+        SWSS_LOG_INFO("Delete labeled next hop %s",
+                        m_nh_key.to_string().c_str());
+        gNeighOrch->removeMplsNextHop(m_nh_key);
+    }
+}
+
+/*
+ * Purpose:     Constructor.
+ *
+ * Description: Initialize the group's members based on the next hop group key.
+ *
+ * Params:      IN  key - The next hop group's key.
+ *
+ * Returns:     Nothing.
+ */
+NextHopGroup::NextHopGroup(const NextHopGroupKey& key) :
+    m_key(key),
+    m_id(SAI_NULL_OBJECT_ID),
+    m_is_temp(false)
+{
+    SWSS_LOG_ENTER();
+    SWSS_LOG_INFO("Creating next hop group %s", m_key.to_string().c_str());
+
+    /* Parse the key and create the members. */
+    for (const auto& it : m_key.getNextHops())
+    {
+        SWSS_LOG_INFO("Adding next hop %s to the group",
+                        it.to_string().c_str());
+        m_members.emplace(it, NextHopGroupMember(it));
+    }
+}
+
+/*
+ * Purpose:     Move constructor.
+ *
+ * Description: Initialize the members by doing member-wise move construct.
+ *
+ * Params:      IN  nhg - The rvalue object to initialize from.
+ *
+ * Returns:     Nothing.
+ */
+NextHopGroup::NextHopGroup(NextHopGroup&& nhg) :
+    m_key(std::move(nhg.m_key)),
+    m_id(std::move(nhg.m_id)),
+    m_members(std::move(nhg.m_members)),
+    m_is_temp(nhg.m_is_temp)
+{
+    SWSS_LOG_ENTER();
+
+    /* Invalidate the rvalue SAI ID. */
+    nhg.m_id = SAI_NULL_OBJECT_ID;
+}
+
+/*
+ * Purpose:     Move assignment operator.
+ *
+ * Description: Perform member-wise swap.
+ *
+ * Params:      IN  nhg - The rvalue object to swap with.
+ *
+ * Returns:     Referene to this object.
+ */
+NextHopGroup& NextHopGroup::operator=(NextHopGroup&& nhg)
+{
+    SWSS_LOG_ENTER();
+
+    std::swap(m_key, nhg.m_key);
+    std::swap(m_id, nhg.m_id);
+    std::swap(m_members, nhg.m_members);
+    m_is_temp = nhg.m_is_temp;
+
+    return *this;
+}
+
+/*
+ * Purpose:     Sync a next hop group.
+ *
+ * Description: Fill in the NHG ID.  If the group contains only one NH, this ID
+ *              will be the SAI ID of the next hop that NeighOrch owns.  If it
+ *              has more than one NH, create a group over the SAI API and then
+ *              add it's members.
+ *
+ * Params:      None.
+ *
+ * Returns:     true, if the operation was successful;
+ *              false, otherwise.
+ */
+bool NextHopGroup::sync()
+{
+    SWSS_LOG_ENTER();
+
+    /* If the group is already synced, exit. */
+    if (isSynced())
+    {
+        SWSS_LOG_INFO("Group %s is already synced", to_string().c_str());
+        return true;
+    }
+
+    /*
+     * If the group has only one member, the group ID will be the member's NH
+     * ID.
+     */
+    if (m_members.size() == 1)
+    {
+        const NextHopGroupMember& nhgm = m_members.begin()->second;
+
+        if (nhgm.getNhId() == SAI_NULL_OBJECT_ID)
```

Comment:
save NHID in a variable to avoid calling twice.

Reply:
Done